Fix the failed to mv error on python s2i

anmolbabu opened the pull request:
The error happens when the local component source contains
a directory by name s2i. This fix clears the $APP_ROOT of
all files and folders before the assemble script copies
over sources from $destination dir to $app-root so that
mv from assemble script does not compile that the folder
already exists

fixes https://github.com/redhat-developer/odo/issues/1054
Signed-off-by: anmolbabu <anmolbudugutta@gmail.com>

anmolbabu commented:
This breaks watch :(

cdrage commented:
Hey @anmolbabu can you please rebase to master? Tests had broken before.. See #11

anmolbabu commented:
Working approach for source deployments :

1. Use backup-dir(/opt/app-root-src-backup) to backup source in $ODO_S2I_SRC_BIN_PATH..
2. rm -fr contents of $APP_ROOT/src so that assemble does not complain error in https://github.com/redhat-developer/odo/issues/1054
3. Trigger assemble script
4. Copy back source from /opt/app-root-src-backup to $APP_ROOT

This way, the sources from $ODO_S2I_SRC_BIN_PATH are backed up before being moved to $APP_ROOT from which the sources will be deleted to void error in https://github.com/redhat-developer/odo/issues/1054 before the same being movede over from $ODO_S2I_SRC_BIN_PATH(which empties it for the next iteration of watch)

Verification of the method being fool proof for every runtime and different combinations of watch and push triggered for binary and local source

cdrage commented:
@anmolbabu 

It looks like the tests are working as intended (I see that it failed in two sections..)

Make sure you look at @kadel 's PR too as he restores from backup so it may conflict with regards to your PR.

anmolbabu commented on assemble-and-restart:
```diff
@@ -2,17 +2,35 @@
+# Backup contents of source/bin directory to source backup in case of component created using source
+if [ -n  "$ODO_SRC_BACKUP_DIR" ]; then
+    mkdir -p ${ODO_SRC_BACKUP_DIR}
+    cp -rf ${ODO_S2I_SRC_BIN_PATH}/src/* ${ODO_SRC_BACKUP_DIR}/
```
use `.` instead of `*` to even copy files and folders starting with `.`

anmolbabu replied:
Source: @kadel

anmolbabu commented on assemble-and-restart:
```diff
@@ -2,17 +2,35 @@
+# Backup contents of source/bin directory to source backup in case of component created using source
+if [ -n  "$ODO_SRC_BACKUP_DIR" ]; then
+    mkdir -p ${ODO_SRC_BACKUP_DIR}
+    cp -rf ${ODO_S2I_SRC_BIN_PATH}/src/* ${ODO_SRC_BACKUP_DIR}/
```
Replace all `cp` with `rsync`

anmolbabu replied:
Source @kadel

anmolbabu commented:
This introduced a new issue with wildfly as in :+1: 
```
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 1.599 s
[INFO] Finished at: 2018-12-20T12:38:35Z
[INFO] ------------------------------------------------------------------------
[WARNING] The requested profile "openshift" could not be activated because it does not exist.
Moving built war files into /wildfly/standalone/deployments for later deployment...
Moving all war artifacts from /opt/app-root/src/target directory into /wildfly/standalone/deployments for later deployment...
'/opt/app-root/src/target/ROOT.war' -> '/wildfly/standalone/deployments/ROOT.war'
removed '/opt/app-root/src/target/ROOT.war'
Moving all ear artifacts from /opt/app-root/src/target directory into /wildfly/standalone/deployments for later deployment...
Moving all rar artifacts from /opt/app-root/src/target directory into /wildfly/standalone/deployments for later deployment...
Moving all jar artifacts from /opt/app-root/src/target directory into /wildfly/standalone/deployments for later deployment...
mkdir: cannot create directory '/wildfly/provided_modules': File exists
```

anmolbabu committed:
Fix the failed to mv error on python s2i

The error happens when the local component source contains
a directory by name s2i. This fix clears the $APP_ROOT of
all files and folders before the assemble script copies
over sources from $destination dir to $app-root so that
mv from assemble script does not compile that the folder
already exists

fixes redhat-developer/odo#1054
Signed-off-by: anmolbabu <anmolbudugutta@gmail.com>

anmolbabu commented:
> @anmolbabu
> 
> It looks like the tests are working as intended (I see that it failed in two sections..)
> 
> Make sure you look at @kadel 's PR too as he restores from backup so it may conflict with regards to your PR.

@cdrage Ack 
Fortunately those lines remained untouched in my PR ;) So hopefully they shouldn't conflict

geoand reviewed (approved):
Tested and works!

geoand commented on assemble-and-restart:
```diff
@@ -2,6 +2,32 @@
+        fi
+    done
+    # In case of java clear off the target directory
+    rm -fr ${ODO_S2I_SRC_BIN_PATH}/src/target
```
This is needed because java s2i does not delete the previous compilation output. If this weren't added, then deleted source files would not be reflected in the compilation output

kadel replied:
This is java specific, I  would like to avoid working with paths that are just for valid just for one type of builder images for now. We need a better way to handle that.

What if my project source that is in a different language contains `src/target` path for some reason?

geoand replied:
Unfortunately, if we don't delete this directory, watch + delete is totally broken for Java.
Perhaps we should wrap the delete call in an if statement that checks if the image is a Java image (by testing for the existence of the mvn binary)?

anmolbabu replied:
Yeah I added the conditional check now
@kadel @geoand can you have a look now if it is ok?

geoand replied:
@anmolbabu I don't think that using `"$ODO_S2I_BUILDER_IMG" == "redhat-openjdk-18/openjdk18-openshift"` is correct, because the image builder could have a different name, right?
I personally think that testing for the image capabilities (meaning testing if it has the `mvn` binary) is the correct approach. 
An alternative could be perhaps to check for the existence of `pom.xml` in the pushed source code

anmolbabu replied:
@geoand IMO, this fix is very specific to the way openjdk18 image redhat-openjdk-18/openjdk18-openshift works but java has ways to mitigate it like for example using `mvn clean package` which any other java image might be handling so, we need to check this image in particular?

geoand replied:
The only other Java builder we use at the moment is the widlfly one and I haven't checked to see if it uses `mvn clean`.
My issue is basically with the name - are you sure the builder image name that is injected as an env var will always be the same no matter how I name it in my cluster?

anmolbabu replied:
Yeah this is the name of docker image(not the name it is imported as for ex: I have imported it as openjdk18 like in tests) corresponding to s2i image...
So it is likely not to change

geoand replied:
In that case, I'm fine with the change as long as we check and see if widlfly needs the same kind of handling :)

kadel replied:
What is exactly the issue that this fixes? I have tried a few Java examples and with this commented out and haven't any issues. Except hitting  https://github.com/redhat-developer/odo/issues/1117 again.

geoand replied:
@kadel To see the problem this fixes you need to do the following:

Use the updated source example from this PR.

```
odo create ... 
odo push
odo watch
```

Then delete one of the Java classes. Once the watch picks up the delete, the build should fail (since of the required sources no longer exists).
If you remove this line, then the build does not fail because it retains the previous compiled file

anmolbabu committed:
Remove "target" folder allowing its fresh re-generation if java

Signed-off-by: anmolbabu <anmolbudugutta@gmail.com>

jorgemoralespou commented:
Don't know how to test, but want to make sure this would work:

When doing `odo push` and it's building, hit CTRL-C and break the execution of `odo push`. In current release, somehow backup dir is preserved (or saved, don't remember which) and further builds didn't work.

kadel commented on assemble-and-restart:
```diff
@@ -2,6 +2,34 @@
+    # Clear all those dirs in WorkingDir that exist in DestinationDir so that when assemble script moves sources from
+    # DestinationDir to WorkingDir, it sees the directory clean and doesn't complain:
+    # https://github.com/redhat-developer/odo/issues/1054
+    for file in `ls -a ${ODO_S2I_SRC_BIN_PATH}/src/`
```
you can do `ls -A`  and then you don't need to check for `.` and `..` 

But problem is that this aproach with `for` and `ls` is not going to work properly when there is a filename that contains space!

kadel replied:
it would be good to have case with filename with space in our tests

anmolbabu replied:
Ack

cdrage commented on assemble-and-restart:
```diff
@@ -2,6 +2,34 @@
 set -x
 set -eo pipefail
 
+# If WorkingDir is injected as an env from odo and destination path is not equal to WorkingDir injected by odo
```
This sentence is confusing..

anmolbabu replied:
Ack I will fix it

cdrage commented on assemble-and-restart:
```diff
@@ -2,6 +2,34 @@
+        if [ ! -d ${ODO_SRC_BACKUP_DIR} ]; then
+            mkdir -p ${ODO_SRC_BACKUP_DIR}/src
+        fi
+        # Backup the sources in DestinationDir to ODO_SRC_BACKUP_DIR bcoz the assemble script for some s2i images
```
bcoz = because

anmolbabu replied:
Ack

cdrage commented on assemble-and-restart:
```diff
@@ -2,6 +2,34 @@
+            rm -fr $ODO_S2I_WORKING_DIR/$file
+        fi
+    done
+    # In case of java clear off the target directory
```
This is oddly specific... is there any way to make this more generic or do we actually have to hard-code in openjdk support?

This  will be a future problem when newer versions of openjdk is released in the future of if the image name changes

geoand replied:
Clearing off the output directory should not cause any problems, even if the builder image does this on it's own

geoand replied:
See: https://github.com/redhat-developer/odo-supervisord-image/pull/10#discussion_r247052423

cdrage reviewed (changes requested):
Hey @anmolbabu if you can organize the code a bit that'd be great. Add some newlines between code that's different! :+1:

anmolbabu commented:
> Don't know how to test, but want to make sure this would work:
> 
> When doing `odo push` and it's building, hit CTRL-C and break the execution of `odo push`. In current release, somehow backup dir is preserved (or saved, don't remember which) and further builds didn't work.

@jorgemoralespou I would say not in the scope of this PR
Do you mean hot Ctrl+c should break even an already executing assemble script?
odo doesn't intercept SIGINT at the moment I am not sure if there's a way to do it in Go(I have seen it in python though) so may be I can write a new PR for that bcoz I am really not sure how the SIGINT even if caught, can even be communicated to the component pod where the assemble is under execution

anmolbabu committed:
Incoporate @kadel and @cdrage comments

Signed-off-by: anmolbabu <anmolbudugutta@gmail.com>

kadel commented on assemble-and-restart:
```diff
@@ -2,6 +2,49 @@
+    IFS=$b_IFS
+
+    # In case of java clear off the target directory
+    if [ -n "$ODO_S2I_BUILDER_IMG" ] && [ "$ODO_S2I_BUILDER_IMG" == "redhat-openjdk-18/openjdk18-openshift" ]; then
```
@anmolbabu please add `TODO` comment so we don't forget that this is a temporary hack and we need to find a proper solution for this.

anmolbabu replied:
Ack

kadel commented:
> @jorgemoralespou I would say not in the scope of this PR

This PR is already a significant change (because it is tight to https://github.com/redhat-developer/odo/pull/1125) Let's fix this right after we are done with this.

> Do you mean hot Ctrl+c should break even an already executing assemble script?
> odo doesn't intercept SIGINT at the moment I am not sure if there's a way to do it in Go(I have seen it in python though) so may be I can write a new PR for that bcoz I am really not sure how the SIGINT even if caught, can even be communicated to the component pod where the assemble is under execution

We need to fix it, this was not an issue before we started doing all this backup and restore charades.
Odo should never leave component in a broken state where the only way to fix it is to delete the whole component.

kadel commented:
@jorgemoralespou I just compiled odo with https://github.com/redhat-developer/odo/pull/1125 that uses image from this PR to make it easy to test this anywhere. 

You can download it here [odo.zip](https://github.com/redhat-developer/odo-supervisord-image/files/2768877/odo.zip) ( It is compiled just for MacOS)

cdrage reviewed (approved):
Did some light testing, this LGTM :+1: code + comments + implementation.

anmolbabu commented:
@kadel All comments incoporated please review

metacosm commented:
I don't feel qualified to review this PR, sorry.

anmolbabu committed:
Enhance nodejs builds during re-used subsequent builds

Signed-off-by: anmolbabu <anmolbudugutta@gmail.com>

anmolbabu commented:
Seems for some reason, the wildfly source watch consistently fails with e2e in odo repo....

anmolbabu committed:
Refactoring /opt/app-root/backup to ODO_DEPLOYMENT_BACKUP_DIR

Signed-off-by: anmolbabu <anmolbudugutta@gmail.com>

jorgemoralespou commented on run:
```diff
@@ -4,10 +4,10 @@ set -eo pipefail
-if [ -d /opt/app-root/backup ] && [ -n "$ODO_S2I_DEPLOYMENT_DIR" ]; then
-    cp -rf /opt/app-root/backup/. ${ODO_S2I_DEPLOYMENT_DIR}/
+if [ -d ${ODO_DEPLOYMENT_BACKUP_DIR} ] && [ -n "$ODO_S2I_DEPLOYMENT_DIR" ]; then
+    cp -rf ${ODO_DEPLOYMENT_BACKUP_DIR}/. ${ODO_S2I_DEPLOYMENT_DIR}/
```
Why everywhere in assemble-and-restart this has changed to rsync, but here it has remained as cp? same performance?

anmolbabu replied:
My bad :(

anmolbabu replied:
Fixed now

kadel commented:
Tests are failing because the new image is not compatible with current odo. (It requires https://github.com/redhat-developer/odo/pull/1125 to work properly)

kadel commented on run:
```diff
@@ -4,10 +4,10 @@ set -eo pipefail
-if [ -d /opt/app-root/backup ] && [ -n "$ODO_S2I_DEPLOYMENT_DIR" ]; then
-    cp -rf /opt/app-root/backup/. ${ODO_S2I_DEPLOYMENT_DIR}/
+if [ -d ${ODO_DEPLOYMENT_BACKUP_DIR} ] && [ -n "$ODO_S2I_DEPLOYMENT_DIR" ]; then
+    rsync -r ${ODO_DEPLOYMENT_BACKUP_DIR}/. ${ODO_S2I_DEPLOYMENT_DIR}/
```
it would be better to use `rsync -a` to make sure that it is really identical copy

anmolbabu replied:
Done

kadel commented on assemble-and-restart:
```diff
@@ -13,15 +58,23 @@ else
     fi
-   cp -rf ${ODO_S2I_DEPLOYMENT_DIR}/. /opt/app-root/backup/
+
+   rsync -r ${ODO_S2I_DEPLOYMENT_DIR}/. ${ODO_DEPLOYMENT_BACKUP_DIR}/
```
it would be better to use `rsync -a` to make sure that it is really identical copy

anmolbabu replied:
Done

kadel commented on assemble-and-restart:
```diff
@@ -13,15 +58,23 @@ else
+# After assemble script is run(which delets sources from destination dir by doing mv of them to the WorkingDir) and if only its component created from source,
+# copy back the sources from Source backup dir to DestinationDir for subsequent push of only updates by watch
+if [ ! -z "${ODO_S2I_WORKING_DIR}" ] && [ -n  "$ODO_SRC_BACKUP_DIR" ] && ([ "${ODO_S2I_SRC_BIN_PATH}" != "${ODO_S2I_WORKING_DIR}" ] || [ "${ODO_S2I_DEPLOYMENT_DIR}" != "${ODO_S2I_WORKING_DIR}" ]); then
+    rsync -r ${ODO_SRC_BACKUP_DIR}/src/. ${ODO_S2I_SRC_BIN_PATH}/src/
```
it would be better to use `rsync -a` to make sure that it is really identical copy

anmolbabu replied:
Done

kadel commented on assemble-and-restart:
```diff
@@ -2,6 +2,51 @@
+        # Backup the sources in DestinationDir to ODO_SRC_BACKUP_DIR because the assemble script for some s2i images
+        # moves sources from DestinationDir to WorkingDir thereby deleting it there(DestinationDir). So, back it up
+        # for partial push use cases like watch
+        rsync -r ${ODO_S2I_SRC_BIN_PATH}/src/. ${ODO_SRC_BACKUP_DIR}/src/
```
it would be better to use `rsync -a` to make sure that it is really identical copy

anmolbabu replied:
Done

anmolbabu committed:
Incoporate @kadel comments

Signed-off-by: anmolbabu <anmolbudugutta@gmail.com>
